File: code/Cpp/Day5_NextPermutation.cpp
class Solution {
  public:
    void nextPermutation(vector<int>& arr) {
        
        int n = arr.size();
        int k=-1,l=-1;
        
        for(int i=n-2;i>-1;i--){
            if(arr[i]<arr[i+1]){
                k=i;
                break;
            }
        }
        
        if(k==-1){
            sort(arr.begin(),arr.end());
            return;
        }
        
        
        for(int i=n-1;i>k;i--){
            if(arr[i]>arr[k]) {l=i;
                break;
            }
        }
        
        int temp=arr[k];
        arr[k]=arr[l];
        arr[l]=temp;
        
        reverse(arr.begin() + k + 1, arr.end());
    }
};

File: code/Python/Arrays/Day_5_NextPermutation.py
def nextPermutation(arr):
    n = len(arr)
    
    # Step 1: Find the largest index k such that arr[k] < arr[k + 1]
    k = -1
    for i in range(n - 2, 0, -1):
        if arr[i] < arr[i + 1]:
            k = i
            break
    
    # If no such index exists, the array is the largest permutation
    if k == -1:
        arr.reverse()  #reverse the array to get the lowest permutation
        return arr
    
    # Step 2: Find the largest index l such that arr[k] < arr[l]
    l = -1
    for i in range(n - 1, k, -1):
        if arr[i] > arr[k]:
            l = i
            break
    
    # Step 3: Swap arr[k] and arr[l]
    arr[k], arr[l] = arr[l], arr[k]
    
    # Step 4: Reverse the suffix starting at k + 1 to ensure actual next permutation
    arr[k + 1:] = reversed(arr[k + 1:])
    
    return arr
    
res = nextPermutation([2,4,1,5,7,0])
print(res)
